cmd/fuzzcrypto/std/rsa_test.go

```diff
@@ -0,0 +1,164 @@
+		"3467903426626310123395340254094941045497208049900750380025518552334536945536837294961497712862519984786362199788654739924501424784631315081391467293694361474867825728031147665777546570788493758372218019373",
+		"4597024781409332673052708605078359346966325141767460991205742124888960305710298765592730135879076084498363772408626791576005136245060321874472727132746643162385746062759369754202494417496879741537284589047",
+	},
+}
+
+func newRSAKey(e int, bN, bD, bP, bQ, bR string) *rsa.PrivateKey {
+	fromBase10 := func(base10 string) *big.Int {
+		i, ok := new(big.Int).SetString(base10, 10)
+		if !ok {
+			return big.NewInt(0)
+		}
+		return i
+	}
+	key := &rsa.PrivateKey{
+		PublicKey: rsa.PublicKey{
+			N: fromBase10(bN),
+			E: e,
+		},
+		D:      fromBase10(bD),
+		Primes: []*big.Int{fromBase10(bP), fromBase10(bQ)},
+	}
+	if bR != "" {
+		key.Primes = append(key.Primes, fromBase10(bR))
+	}
+	return key
+}
+
+func FuzzRSAOAEP(f *testing.F) {
+	for _, s := range rsaKeySeeds {
+		f.Add(s.e, s.n, s.d, s.p, s.q, s.r, []byte("testing"), []byte("a label"))
+	}
+	f.Fuzz(func(t *testing.T, e int, bN, bD, bP, bQ, bR string, msg, label []byte) {
+		key := newRSAKey(e, bN, bD, bP, bQ, bR)
+		if key.Validate() != nil {
+			return
+		}
+		enc, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, &key.PublicKey, msg, label)
+		if err != nil {
+			if errors.Is(err, rsa.ErrMessageTooLong) {
+				return
+			}
+			t.Fatal(err)
+		}
+		dec, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, key, enc, label)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !bytes.Equal(dec, []byte(msg)) {
+			t.Errorf("got:%x want:%x", dec, msg)
+		}
+	})
+}
+
+func FuzzRSAPKCS1(f *testing.F) {
+	for _, s := range rsaKeySeeds {
+		f.Add(s.e, s.n, s.d, s.p, s.q, s.r, []byte("testing"))
+	}
+	f.Fuzz(func(t *testing.T, e int, bN, bD, bP, bQ, bR string, msg []byte) {
+		key := newRSAKey(e, bN, bD, bP, bQ, bR)
+		if key.Validate() != nil {
+			return
+		}
+		enc, err := rsa.EncryptPKCS1v15(rand.Reader, &key.PublicKey, msg)
+		if err != nil {
+			if errors.Is(err, rsa.ErrMessageTooLong) {
+				return
+			}
+			t.Fatal(err)
+		}
+		dec, err := rsa.DecryptPKCS1v15(rand.Reader, key, enc)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !bytes.Equal(dec, []byte(msg)) {
+			t.Errorf("got:%x want:%x", dec, msg)
+		}
+	})
+}
+
+func FuzzRSASignPSS(f *testing.F) {
+	for _, s := range rsaKeySeeds {
+		f.Add(s.e, s.n, s.d, s.p, s.q, s.r, []byte("testing"))
+	}
+	opts := rsa.PSSOptions{SaltLength: rsa.PSSSaltLengthEqualsHash}
+	f.Fuzz(func(t *testing.T, e int, bN, bD, bP, bQ, bR string, msg []byte) {
+		key := newRSAKey(e, bN, bD, bP, bQ, bR)
+		if key.Validate() != nil {
+			return
+		}
+		hashed := sha256.Sum256(msg)
+		sig, err := rsa.SignPSS(rand.Reader, key, crypto.SHA256, hashed[:], &opts)
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = rsa.VerifyPSS(&key.PublicKey, crypto.SHA256, hashed[:], sig, &opts)
+		if err != nil {
+			t.Error(err)
+		}
+		hashed[0] ^= 0xff
```

Comment:
In theory we could improve the fuzzing domain by having the fuzzer pass us non-zero `[sha256.Size]byte` arrays to xor the entire checksum against, right? (Well, don't necessarily have to use xor for that, but I think you get what I mean. 😄)

I don't know how this extra domain would be *useful* to fuzz, though, personally. (What would it be, some kind of hash collision in the underlying library? And the extra time it would take to run each iteration might mean we don't catch some other issue in the parts of the domain that matter....)

Reply:
RSA sign tests already have too many parameters, it takes almost 1 minute to start converging! so I'm reluctant to another slice input. Additionally, all the verification logic is outside the Go stack, so the fuzzing engine won't learn anything by xoring bits, the xor array would be completely random.

On the other hand, your point is valid, the hash collision coverage is low with the current setup. I'll implement a compromise: add a new `uint8` input which will be used to decide which byte of the hash to xor and with which value. At least we will not be xoring always the first byte with `0xff`.